src/hotspot/share/opto/node.hpp

```diff
@@ -1392,6 +1392,9 @@ class DUIterator_Fast : public DUIterator_Common {
   DUIterator_Fast()
     { /*initialize to garbage*/         debug_only(_vdui = false); }
 
+  DUIterator_Fast(const DUIterator_Fast& that)
+    { _outp = that._outp;               debug_only(reset(that)); }
```

Comment:
`reset` tests `_vdui`, but nothing here has set it, so it's uninitialized and that test wil be UB.

I'm also not sure why it's okay for `operator=` to use whatever the current value of `_vdui` might be; that could leave `_last` as the old value rather than refreshing from `that`, which seems wrong.  This is aabout pre-existing code that looks questionable to me.

Reply:
I suppose the constructor would be invoked before the copy assignment operator. That is `_vdui` gets initialized already in the ctor `DUIterator_Fast()` for `operator=` case. Right?

Yes. For our newly-added copy constructor for `DUIterator_Fast`, we should initialize `_vdui` as `false`. It may be defined as below.
```
DUIterator_Fast(const DUIterator_Fast& that)
    { _outp = that._outp;               debug_only(_vdui = false; reset(that)); }
```

Reply:
That's true on the first assignment of `t2`. But what if `t2` is reassigned
to some other iterator. That assignment sees `_vdui` true, and keeps the old
value of `_last` rather than updating the value from that other iterator. Is
that really correct? It certainly seems strange. I'm trying to find someone
who understands this code to get involved, but holidays.

Reply:
Thanks for your explanation. Yes, you're right. I didn't realize the re-assignment scenario. 

However, I noticed that there does NOT exist such re-assignment cases. 
Simply grep `= DUIterator_Last(` and `= DUIterator_Fast(`.
Is there anything I missed?

Reply:
@vnkozlov I was wondering if you could take a look at this? We're not sure whether 'operator=' is problematic or not. Thanks.

Reply:
I manually checked the usages of assignment operators for class DUIterator, DUIterator_Fast and DUIterator_Last. (Simply grep the class names in the source code and check the context). 

I found there exist only a couple of re-assignment usages. However, I guess kind of reset operations are conducted in these sites, and I suspect  the implementation of `operator=` might be good.

As I examined, only the following 3 sites are found where re-assignment happens
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/cfgnode.cpp#L565
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/phaseX.cpp#L1878
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/split_if.cpp#L452

Please take `cfgnode.cpp` as an example. `j` is first assigned at line 568 and it would be assigned again if flag `progress` is true. However, I suppose `j` gets reset at line 578 in such case.

It might be incorrect if I missed something.

Reply:
Code is correct. _last should not be updated with additional assignments which updates only node's information.

Reply:
Thanks a lot for your explanation @vnkozlov .